service/history/workflowResetor.go

```diff
@@ -558,3 +594,212 @@ func validateLastBatchOfReset(lastBatch []*workflow.HistoryEvent) error {
 		Message: fmt.Sprintf("wrong DecisionFinishEventId, previous batch doesn't include EventTypeDecisionTaskStarted, lastFirstEventId: %v", firstEvent.GetEventId()),
 	}
 }
+
+func validateResetReplicationTask(request *h.ReplicateEventsRequest) (*workflow.DecisionTaskFailedEventAttributes, error) {
+	historyAfterReset := request.History.Events
+	if len(historyAfterReset) == 0 || historyAfterReset[0].GetEventType() != workflow.EventTypeDecisionTaskFailed {
+		return nil, errUnknownReplicationTask
+	}
+	firstEvent := historyAfterReset[0]
+	if firstEvent.DecisionTaskFailedEventAttributes.GetCause() != workflow.DecisionTaskFailedCauseResetWorkflow {
+		return nil, errUnknownReplicationTask
+	}
+	attr := firstEvent.DecisionTaskFailedEventAttributes
+	if attr.GetNewRunId() != request.GetWorkflowExecution().GetRunId() {
+		return nil, errUnknownReplicationTask
+	}
+	return attr, nil
+}
+
+func (w *workflowResetorImpl) ApplyResetEvent(ctx context.Context, request *h.ReplicateEventsRequest, domainID, workflowID, currentRunID string) (retError error) {
+	var currContext workflowExecutionContext
+	var forkMutableState, currMutableState, newMsBuilder mutableState
+	var newRunTransferTasks, newRunTimerTasks []persistence.Task
+
+	forkAttr, retError := validateResetReplicationTask(request)
+	historyAfterReset := request.History.Events
+	lastEvent := historyAfterReset[len(historyAfterReset)-1]
+	decisionFinishEventID := historyAfterReset[0].GetEventId()
+	if retError != nil {
+		return
+	}
+	forkExecution := workflow.WorkflowExecution{
+		WorkflowId: common.StringPtr(workflowID),
+		RunId:      common.StringPtr(forkAttr.GetForkRunId()),
+	}
+
+	forkContext, forkRelease, retError := w.eng.historyCache.getOrCreateWorkflowExecutionWithTimeout(ctx, domainID, forkExecution)
+	if retError != nil {
+		return retError
+	}
+	defer func() { forkRelease(retError) }()
+	forkMutableState, retError = forkContext.loadWorkflowExecution()
+	if retError != nil {
+		return
+	}
+	if forkMutableState.GetNextEventID() < decisionFinishEventID {
+		// re-replicate the whole new run
+		return newRetryTaskErrorWithHint(ErrWorkflowNotFoundMsg, domainID, workflowID, forkAttr.GetNewRunId(), common.FirstEventID)
+	}
+
+	if currentRunID == forkAttr.GetForkRunId() {
+		currMutableState = forkMutableState
+		currContext = forkContext
+	} else {
+		var currRelease releaseWorkflowExecutionFunc
+		currExecution := workflow.WorkflowExecution{
+			WorkflowId: forkExecution.WorkflowId,
+			RunId:      common.StringPtr(currentRunID),
+		}
+		currContext, currRelease, retError = w.eng.historyCache.getOrCreateWorkflowExecutionWithTimeout(ctx, domainID, currExecution)
```

Comment:
maybe this should be currErr?

line 631 should probably be forkErr?

Reply:
I am OK to do that. I guess the benefit of doing that is with currErr, we can avoid releasing forkMutableState without clear the cache. I don't quite know what do you want to achieve to use forkErr. But that's also fine to be consistent.